cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java

```diff
@@ -393,29 +442,57 @@ private JSONObject normalizeRemotePrincipal(final OpenScienceFrameworkCredential
             throws ParserConfigurationException, TransformerException {
         final DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
         final DocumentBuilder builder = factory.newDocumentBuilder();
-
         final Document document = builder.newDocument();
         final Element rootElement = document.createElement("auth");
-        document.appendChild(rootElement);
 
-        for (final String key : credential.getAuthenticationHeaders().keySet()) {
+        document.appendChild(rootElement);
+        for (final String key : credential.getDelegationAttributes().keySet()) {
             final Element attribute = document.createElement("attribute");
             attribute.setAttribute("name", key);
-            attribute.setAttribute("value", credential.getAuthenticationHeaders().get(key));
+            attribute.setAttribute("value", credential.getDelegationAttributes().get(key));
             rootElement.appendChild(attribute);
         }
 
         // run the auth document through the transformer
         final DOMSource source = new DOMSource(document);
         final StringWriter writer = new StringWriter();
         final StreamResult result = new StreamResult(writer);
-
         this.institutionsAuthTransformer.transform(source, result);
 
         // convert transformed xml to json
         return XML.toJSONObject(writer.getBuffer().toString());
     }
 
+    /**
+     * Update `credential` with remote principal authenticated by external CAS.
+     *
+     * @param credential the credential whose authenticationHeaders is going to be udpated
+     * @param principal the principal created by external CAS authentication
+     * @return the updated credential
+     * @throws AccountException RemoteUserFailedLoginException
+     */
+    private OpenScienceFrameworkCredential constructCredentialFromRemotePrincipal(
```

Comment:
Update `credential.delegationAttributes`:
- `Delegation-Protocol`: a control variable for XSLT to choose which type of transformation to apply.
- `Cas-Identity-Provider`: use `clientName` which is identical to `institutionId`
- `attributes` in `principal`
  - OK State University
    - `mail` is mapped to `username`
    - `sn` is mapped to `familyName`
    - `givenName` is mapped to `givenName`
- Refer to the `institutions-auth.xsl` I airdropped to you.